src/monodroid/jni/monodroid-glue.c

```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
+	mono_bool success = add_reference_jobject (env, handle, reffed_handle);
+
+	// Flag MonoObjects so they can be cleared in gc_cleanup_after_java_collection.
+	// Java temporaries do not need this because the entire GCUserPeer is discarded.
+	if (success && target.is_mono_object) {
+		int ref_val = 1;
+		mono.mono_field_set_value (target.obj, bridge_info->refs_added, &ref_val);
+	}
+
+#if DEBUG
+	if (gc_spew_enabled) {
+		char *description = describe_target (target),
+			 *reffed_description = describe_target (reffed_target);
+
+		if (success)
+			log_warn (LOG_GC, "Added reference for %s to %s", description, reffed_description);
+		else
+			log_error (LOG_GC, "Missing monodroidAddReference method for %s", description);
+
+		free (description);
+		free (reffed_description);
+	}
+#endif
+
+	return success;
+}
+
+// Create a target
+AddReferenceTarget mono_object_target (MonoObject *obj) {
+	AddReferenceTarget result;
+	result.is_mono_object = TRUE;
+	result.obj = obj;
+	return result;
+}
+
+// Create a target
+AddReferenceTarget jobject_target (jobject jobj) {
+	AddReferenceTarget result;
+	result.is_mono_object = FALSE;
+	result.jobj = jobj;
+	return result;
+}
+
+// Extract the root target for an SCC. If the SCC has bridged objects, this is the first object. If not, it's stored in temporary_peers.
+// Must pass in the JNI information necessary to interact with temporary_peers.
+static AddReferenceTarget scc_target (MonoGCBridgeSCC **sccs, int idx, JNIEnv *env, jobject temporary_peers, jmethodID arraylist_get, int *temporary_peer_indices)
+{
+	MonoGCBridgeSCC *scc = sccs [idx];
+	if (scc->num_objs > 0)
+		return mono_object_target (scc->objs [0]);
+
+	jobject jobj = (*env)->CallObjectMethod (env, temporary_peers, arraylist_get, temporary_peer_indices [idx]);
+	return jobject_target (jobj);
+}
+
+// Must call this on any AddReferenceTarget returned by scc_target once done with it
+static void release_scc_target (JNIEnv *env, AddReferenceTarget target) {
+	if (!target.is_mono_object)
+		(*env)->DeleteLocalRef (env, target.jobj);
+}
+
+// Add a reference between objects if both are already known to be MonoObjects which are user peers
+static mono_bool
+add_reference_mono_object (JNIEnv *env, MonoObject *obj, MonoObject *reffed_obj)
+{
+	return add_reference (env, mono_object_target (obj), mono_object_target (reffed_obj));
+}
+
+// Clear a local reference which might be NULL // FIXME: Is it safe to just call DeleteLocalRef with NULL?
+static void
+delete_java_lref (JNIEnv *env, jobject jobj)
+{
+	if (jobj)
+		(*env)->DeleteLocalRef (env, jobj);
+}
+
 static void
 gc_prepare_for_java_collection (JNIEnv *env, int num_sccs, MonoGCBridgeSCC **sccs, int num_xrefs, MonoGCBridgeXRef *xrefs)
 {
-	MonoGCBridgeSCC *scc;
-	int i, j, ref_val;
+	/* Some SCCs might have no IGCUserPeers associated with them, so we must create one */
+	jobject arraylist = NULL, gcuserpeer = NULL;
+	jmethodID gcuserpeer_constructor = NULL, arraylist_get = NULL, arraylist_add = NULL;
```

Comment:
> That's acceptable, then

It's only two more, so yes.

> Are class/methodid lookups expensive

Expensive enough that we want to avoid them when possible. See also: https://bugzilla.xamarin.com/show_bug.cgi?id=32126

>  I notice we're calling GetMethodID on every object we add a reference for, currently

Indeed! This is also something [addressed in Java.Interop](https://github.com/xamarin/java.interop/blob/9fbade1/src/java-interop/java-interop-gc-bridge-mono.c#L845-L859), and we can worry about it in the eventual Java.Interop/xamarin-android GC bridge reunion.